### PR TITLE
version: add RootlessKit, slirp4netns, and VPNKit version

### PR DESCRIPTION
**Pull request by AkihiroSuda**
**- What I did**
Add RootlesKit info to the Version API.

**- How to verify it**
```console
$ docker --context=rootless version
Client:
 Version:           unknown-version
 API version:       1.41
 Go version:        go1.16.3
 Git commit:        unknown-commit
 Built:             unknown-buildtime
 OS/Arch:           linux/amd64
 Context:           rootless
 Experimental:      true

Server:
 Engine:
  Version:          library-import
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.3
  Git commit:       library-import
  Built:            library-import
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.5.0
  GitCommit:        8c906ff108ac28da23f69cc7b74f8e7a470d1df0
 runc:
  Version:          1.0.0-rc93+dev
  GitCommit:        c73d766ee6e1a24517545b0a71fe78d5f78d0449
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
 rootlesskit:
  Version:          0.14.2
  ApiVersion:       1.1.1
  NetworkDriver:    slirp4netns
  PortDriver:       builtin
  StateDir:         /tmp/rootlesskit158605632
 slirp4netns:
  Version:          1.1.9
  GitCommit:        4e37ea557562e0d7a64dc636eff156f64927335e
```

**- Description for the changelog**

version: add RootlessKit, slirp4netns, and VPNKit version

**- A picture of a cute animal (not mandatory but encouraged)**
:penguin:

### CODE REVIEW

File: api/types/types.go [hunk 1]
```diff
@@ -212,6 +213,7 @@ type Info struct {
 	SecurityOptions     []string
 	ProductLicense      string               `json:",omitempty"`
 	DefaultAddressPools []NetworkAddressPool `json:",omitempty"`
+	RootlessKit         *rootlesskitapi.Info `json:",omitempty"`
```
Comment (AkihiroSuda) [inline api/types/types.go:216]: Curious, should this be json.RawMessage?

Comment (AkihiroSuda) [reply api/types/types.go]: updated to use `json.RawMessage`

File: rootless/rootless.go [hunk 2]
```diff
@@ -23,3 +31,45 @@ func RunningWithRootlessKit() bool {
...
+	}
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			DialContext: func(ctx context.Context, _, _ string) (net.Conn, error) {
+				var d net.Dialer
+				return d.DialContext(ctx, "unix", apiSock)
+			},
+		},
+	}
+	req, err := http.NewRequest("GET", "http://rootlesskit/v1/info", nil)
```
Comment (thaJeztah) [inline rootless/rootless.go:56]: This feels like we're making an impromptu API client; could we have a basic API client in the rootlesskit repository that we can vendor and consume?

File: rootless/rootless.go [hunk 3]
```diff
@@ -23,3 +31,45 @@ func RunningWithRootlessKit() bool {
 	})
 	return runningWithRootlessKit
 }
+
+// GetRootlessKitInfo returns the Info object of RootlessKit REST API: https://github.com/rootless-containers/rootlesskit/blob/master/pkg/api/openapi.yaml
+//
+// The object is returned as a json.RawMessage, without unmarshalling into a struct of github.com/rootless-containers/rootlesskit/pkg/api.Info ,
+// so as to retain all the fields that are unknown to github.com/rootless-containers/rootlesskit/pkg/api.Info .
+func GetRootlessKitInfo(ctx context.Context) (json.RawMessage, error) {
+	stateDir := os.Getenv("ROOTLESSKIT_STATE_DIR")
```
Comment (thaJeztah) [inline rootless/rootless.go:40]: I realize we may not do this currently for the other binaries, but the `/info` endpoint is already heavy on doing  a _lot_ of things; would it make sense to use a `sync.Once`?

Or are there situations where the rootlesskit binary can be updated _without_ the daemon itself being restarted? (in that case we _could_ use a "dirty" trick to reset the `sync.Once` on SIGHUP (e.g.); I did that for a test in the CLI in;
https://github.com/docker/cli/blob/a32cd16160f1b41c1c4ae7bee4dac929d1484e59/cli/config/config.go#L27

```go
var (
	initConfigDir = new(sync.Once)
```

https://github.com/docker/cli/blob/a32cd16160f1b41c1c4ae7bee4dac929d1484e59/cli/config/config.go#L45-L50

```go
// resetConfigDir is used in testing to reset the "configDir" package variable
// and its sync.Once to force re-lookup between tests.
func resetConfigDir() {
	configDir = ""
	initConfigDir = new(sync.Once)
}
```

File: api/types/types.go [hunk 4]
```diff
@@ -212,7 +213,10 @@ type Info struct {
 	SecurityOptions     []string
 	ProductLicense      string               `json:",omitempty"`
 	DefaultAddressPools []NetworkAddressPool `json:",omitempty"`
-	Warnings            []string
+	// RootlessKit is unmarshallable into a struct of github.com/rootless-containers/rootlesskit/pkg/api.Info .
+	// (not unmarshalled by default, so as to propagate all the fields that are unknown to github.com/rootless-containers/rootlesskit/pkg/api.Info .)
+	RootlessKit json.RawMessage `json:",omitempty"`
```
Comment (thaJeztah) [inline api/types/types.go:218]: I'm a bit on the fence on adding this as a new field. We originally added the `ContainerdCommit`, `RuncCommit` and `InitCommit` fields when those versions _had to_ match a specific version/commit (API's and their behavior wasn't really  "stable" yet, and we received various bugreports that turned out to be because non-default versions were installed).

The `RuncCommit` already has shown to be a bit problematic for situations where another OCI runtime is installed (`crun`, e.g.). We fixed those, but it's still a bit odd to have hard-coded `Runc` now containing "non-runc" information.

In short; I would like to sunset those fields, and instead consider using the `Components` information that we currently use in the `/version` endpoint; https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/daemon/info.go#L98-L114


I _think_ most (if not all) of this information is for "informational" purposes, and not a strict "API" contract , so having the information (other than `Version` and `Name`) in a more "free-form" format (just `key/value` information) would probably work, and doing so would keep the API more "stable" (and flexible)

WDYT?

**Review by thaJeztah — CHANGES_REQUESTED** [review]
left some thoughts

File: api/types/types.go (hunk 4 above)
Comment (AkihiroSuda) [reply api/types/types.go:218]: The purpose of this field is to show network driver and port driver info rather than the version, so types.ComponentVersion doesn't seem to fit here IMO.

File: rootless/rootless.go (hunk 3 above)
Comment (AkihiroSuda) [reply rootless/rootless.go:40]: Info may have dynamic info (in future), so sync.Once doesn't seem correct.

File: api/types/types.go (hunk 4 above)
Comment (thaJeztah) [reply api/types/types.go:218]: The `types.ComponentVersion` allows to provide arbitrary other information. The only limitation is that it has a single level key/value pairs, not nested, so relevant information that's nested would have to be slightly changed, but probably would allow for better formatting when presenting in `docker info`, e.g. something like:

```go
Components: []types.ComponentVersion{
    {
        Name:    "rootlesskit",
        Version: "v0.14.2",
        Details: map[string]string{
            "apiVersion":        "1.1.1",
            "stateDir":          "/tmp/rootlesskit648249121",
            "networkDriver":     "slirp4netns",
            "networkDriver.dns": "10.0.2.3, 8.8.8.8",
            "portDriver":        "builtin",
            "portDriver.protos": "tcp, tcp4, tcp6, udp, udp4, udp6",
        },
    },
},
```

This is similar to what we do for storage drivers (only that one uses a slice of `key value` pairs;

Screenshot 2021-05-01 at 14 01 42
Screenshot 2021-05-01 at 14 01 50

File: rootless/rootless.go (hunk 2 above)
Comment (AkihiroSuda) [reply rootless/rootless.go]: updated to import `rootlesskit/pkg/api/client`

File: api/types/types.go (hunk 4 above)
Comment (AkihiroSuda) [reply api/types/types.go]: updated to add rootlesskit version to `/version` endpoint.

Adding `Components`  to `/info` would be another PR.

**Comment by AkihiroSuda** [comment]
@thaJeztah PTAL

File: daemon/info_unix.go [hunk 5]
```diff
@@ -217,6 +218,62 @@ func (daemon *Daemon) fillPlatformVersion(v *types.Version) {
 	} else {
 		logrus.Warnf("failed to retrieve %s version: %s", defaultInitBinary, err)
 	}
+
+	daemon.fillRootlessVersion(v)
+}
+
+func (daemon *Daemon) fillRootlessVersion(v *types.Version) {
+	if !rootless.RunningWithRootlessKit() {
```
Comment (thaJeztah) [inline daemon/info_unix.go:226]: Do you think we should use a `sync.Once` for these? (probably ok for a follow-up)

**Review by thaJeztah — APPROVED** [review]
LGTM

the only thing I'm starting to wonder is if we should have some option to switch between "verbose" and "less verbose" on the CLI (to not print all the components, or only basic information) WDYT?

File: daemon/info_unix.go (hunk 5 above)
Comment (AkihiroSuda) [reply daemon/info_unix.go:226]: RunningWithRootlessKit already has sync.Once: https://github.com/moby/moby/blob/33c332ad1982019cb9250ed0163ee2ed4955f596/rootless/rootless.go#L20

Calling the RootlessKit API should not be `Once`.

**Comment by AkihiroSuda** [comment]
> the only thing I'm starting to wonder is if we should have some option to switch between "verbose" and "less verbose" on the CLI (to not print all the components, or only basic information) WDYT?

SGTM

**Comment by tianon** [comment]
> the only thing I'm starting to wonder is if we should have some option to switch between "verbose" and "less verbose" on the CLI (to not print all the components, or only basic information) WDYT?

+1 - having `docker version` print the full version of several dependent (but separate) components instead of just client/server versions by default is strange IMO :sweat_smile:

**Comment by AkihiroSuda** [comment]
@tianon Does this PR LGTY? :pray:

**Comment by AkihiroSuda** [comment]
rebased

**Comment by AkihiroSuda** [comment]
rebased again

**Comment by AkihiroSuda** [comment]
rebased

**Comment by AkihiroSuda** [comment]
rebased

**Review by tonistiigi — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
rebased 😞

**Comment by AkihiroSuda** [comment]
rebased

**Comment by AkihiroSuda** [comment]
@thaJeztah @tonistiigi Can we merge this?

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by cpuguy83** [comment]
Sorry, looks like there's some conflicts again.

**Comment by AkihiroSuda** [comment]
rebased
